python_modules/dagster/dagster/_core/execution/asset_backfill.py

```diff
@@ -1274,16 +1274,29 @@ def get_canceling_asset_backfill_iteration_data(
         )
 
     failed_subset = AssetGraphSubset.from_asset_partition_set(
-        set(_get_failed_asset_partitions(instance_queryer, backfill_id, asset_graph)),
+        set(
+            _get_failed_asset_partitions(
+                instance_queryer,
+                backfill_id,
+                asset_graph,
+                materialized_subset=updated_materialized_subset,
+            )
+        ),
         asset_graph,
     )
+    # we fetch the failed_subset to get any new assets that have failed and add that to the set of
+    # assets we alerady know failed and their downstreams. However we need to remove any assets in
+    # updated_materialized_subset to account for the case where a run retry successfully
+    # materialized a previlusly failed asset.
+    updated_failed_subset = (
+        asset_backfill_data.failed_and_downstream_subset | failed_subset
+    ) - updated_materialized_subset
     updated_backfill_data = AssetBackfillData(
         target_subset=asset_backfill_data.target_subset,
         latest_storage_id=asset_backfill_data.latest_storage_id,
         requested_runs_for_target_roots=asset_backfill_data.requested_runs_for_target_roots,
         materialized_subset=updated_materialized_subset,
-        failed_and_downstream_subset=asset_backfill_data.failed_and_downstream_subset
-        | failed_subset,
+        failed_and_downstream_subset=updated_failed_subset,
```

Comment:
i'm not super familiar with this code but how is the 'and downstream' part of 'failed_and_downstrema_subset' affected by the retry case? If we add downstream assets as well to the failed list as soon as we detect a partition fails, do we need to remove them if that run is retried? Or are those added at the very end after we have decided that the backfill is totally done and failed?

Reply:
so this specifically is for the case when the backfill has been canceled. 

In the normal code path what we do is:
- find which partitions failed
- do a BFS search to find all downstream partitions of the failed partitions
- set that as the `failed_and_downstream_subset`
- when determining if we can launch a run for a partition, if it is in the `failed_and_downstream_subset` we know we cannot launch the run 

we run this process from scratch on each backfill iteration, so we can handle run retries in the first step by checking if each failed partition has been retried successfully, and if so we remove it from the list of failed partitions. then all the following steps will produce the correct results

for this cancling case it's a bit different. the way it works now is that we take the previous `failed_and_downstream_subset` and just add any partitions that have failed (effectively this just adds the partitions that failed since the last tick). we don't recalculate the downstream subset for any newly failed partitions (i'm not sure why, maybe @clairelin135 knows). 

If we leave the code as is, we could have a case where a failed partition was retried and that retry succeeded since the previous tick. the partition would be in the `failed_and_downstream_subset` from the previous tick, but is now successful. we want to make sure the right status is reported, so we need to remove it from the final set of failed partitions. and we can do that be removing the `materialized_subset` from the `failed_and_downstream_subset`

Another option would be to do the process we do in a normal tick and generate the `failed_and_downstream_subset` from scratch rather than do this additive approach. Not sure why that wasn't done in the first place though, so i didn't make that change